### ISSUE TITLE
twisted.python.test.test_components fails with zope.interface 3.6.4

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5215]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5215|
|Type|defect|
|Created|2011-08-02 18:08:09Z|
|Branch|https://github.com/twisted/twisted/tree/zi264-5215|

zope.interface recently changed how interfaces are hashed:

http://bazaar.launchpad.net/~ztk-steering-group/zope.interface/trunk/revision/136

This changes the behavior of `twisted.python.components` in a way that breaks some usages.  The tests use nested interface definitions with overlapping `__module__`/`__name__` values, which now results in the registry thinking there are duplicates being registered, which it rejects.

  1. It's sort of bad that the tests use nested interface definitions.  It's not a realistic use of interfaces.
  1. It's sort of bad that the tests mutate the global registry.
  1. It's sort of bad that there is state left over after the tests.

<details><summary>Searchable metadata</summary>

```
trac-id__5215 5215
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_zi264_5215 branches/zi264-5215
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1312308489000000 1312308489000000
changetime__1313844848000000 1313844848000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [32414]) Branching to 'zi264-5215'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [32415]) Change the twisted.python.components tests to mutate a scratch registry instead of the global registry

refs [#5215](https://github.com/twisted/twisted/issues/5215)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [32416]) Switch away from nested interface definitions to just using the globally defined interfaces already available.

refs [#5215](https://github.com/twisted/twisted/issues/5215)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [32417]) Add some docstrings; fix one more nested interface definition in _multipleInterfacesForClassOrInterface / test_multipleInterfacesForInterface

refs [#5215](https://github.com/twisted/twisted/issues/5215)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Changed some things!  Seems better now.  [Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/zi264-5215).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @exarkun|
|-|-|

Looks great, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

You're just missing a NEWS fragment.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set status to `closed`|
|-|-|

(In [32468]) Merge zi264-5215 

Author: exarkun
Reviewer: therve
Fixes: [#5215](https://github.com/twisted/twisted/issues/5215)

Change the `twisted.python.components` unit tests to avoid mutating the global
adapter registry and to avoid using nested class definitions as adapters and
interfaces.  Aside from generally being nicer by improving isolation and
repeatability, this lets the unit tests work with zope.interface 2.6.4 which
changes some of the hashing behavior of interface objects.
